include/nbl/asset/utils/IShaderCompiler.h

```diff
@@ -157,32 +161,291 @@ class NBL_API2 IShaderCompiler : public core::IReferenceCounted
+				struct CPUShaderCreationParams {
+					IShader::E_SHADER_STAGE stage;
+					IShader::E_CONTENT_TYPE contentType; //I think this one could be skipped since it's always going to be SPIR-V
+					std::string filepathHint;
+					uint64_t codeByteSize = 0;
+					uint64_t offset = 0; // Offset into the serialized .bin for the Cache where code starts
+				};
+
+				std::string mainFileContents;
+				SCompilerData compilerData;
+				// Keeping this one commented out. Could be useful for lazy-loading of dependencies if it takes up too much memory
+				// uint64_t entryID;
+
+				dependency_container_t dependencies;
+				mutable CPUShaderCreationParams shaderParams;
+				bool serialized = false;
+				mutable core::smart_refctd_ptr<asset::ICPUShader> value = nullptr;
+			};
+
+			inline void insert(SEntry&& entry)
+			{
+				m_container.insert(std::move(entry));
 			}
 
-			virtual IShader::E_CONTENT_TYPE getCodeContentType() const { return IShader::E_CONTENT_TYPE::ECT_UNKNOWN; };
+			// can move to .cpp and have it not inline
+			inline core::smart_refctd_ptr<asset::ICPUShader> find(const SEntry& mainFile, CIncludeFinder* finder)
+			{
+				auto foundRange = m_container.equal_range(mainFile);
+				for (auto& found = foundRange.first; found != foundRange.second; found++)
+				{
+					bool allDependenciesMatch = true;
+					// go through all dependencies
+					for (auto i = 1; i != found->dependencies.size(); i++)
+					{
+						const auto& dependency = found->dependencies[i];
+
+						IIncludeLoader::found_t header;
+						if (dependency.standardInclude)
+							header = finder->getIncludeStandard(dependency.requestingSourceDir, dependency.identifier);
+						else
+							header = finder->getIncludeRelative(dependency.requestingSourceDir, dependency.identifier);
+
+						if (header.hash != dependency.hash || header.contents != dependency.contents)
+						{
+							allDependenciesMatch = false;
+							break;
+						}
+					}
+					if (allDependenciesMatch) {
+						if (!found->value) { // Load shader if not already loaded
+							// If the Cache has no storageBuffer, then it's fresh new (not picked up from serialization): all compiled shaders should be loaded
+							// If codeByteSize is 0, then the Entry found lives in the Cache but not in the serialization, so the Entry was generated in the current runtime:
+							// This means the shader should be loaded
+							assert(storageBuffer && found->shaderParams.codeByteSize != 0);
+							auto code = core::make_smart_refctd_ptr<ICPUBuffer>(found->shaderParams.codeByteSize);
+							memcpy(code->getPointer(), (uint8_t*)storageBuffer->getPointer() + CONTAINER_JSON_SIZE_BYTES + containerJsonSize + found->shaderParams.offset, found->shaderParams.codeByteSize);
+							found->value = core::make_smart_refctd_ptr<ICPUShader>(std::move(code), found->shaderParams.stage, found->shaderParams.contentType, std::move(std::string(found->shaderParams.filepathHint)));
+						}
+						return found->value;
```

Comment:
you do realize the only reason you're getting "double storage" is because you are creating a new `ICPUBuffer` with a default allocator for the `found->value` and copying the data across while holding onto the old serialization CPUBuffer

You could simply just create the ICPUBuffers when deserializing and not hold onto the buffer with serialization data.

Or if you truly wanted no RAM-copies, then you can have a special method to deserialize from an `IFile` where you create your own `CMappedFileCPUBuffer : public ICPUBuffer` that holds onto the source memory mapped `IFile` and returns the offsetted file mapped pointer with `getPointer()`

Reply:
I'll change it to load everything at Cache deserialization then

Reply:
are you going to make a `deserialize` from an `IFile` or you're leaving it as a TODO for someone else?

Reply:
As in, deserialize the whole Cache from an IFile at once, or the one that creates some sort of mapped file buffer and retrieves shaders on demand? I could do the first soon, the other I left for later for now. Do I add a commented out TODO with that method?

Reply:
> Do I add a commented out TODO with that method?

yes
```cpp
/*
// `serializedCache` must be mapped with at least a read flag
static core::smart_refctr_ptr<CCache> deserialize(const system::IFile* serializedCache);
*/
```

Reply:
write this up in a GH issue you'll open about a streaming cache